### ISSUE TITLE
lefthook's location is ignored in the hooks in a monorepo

### ISSUE THREAD

**Issue by shamilovtim** [evilmartians/lefthook#443]
### :wrench: Summary

The lefthook.yml supports the `root:` directive in order to run the hooks from a separate monorepo folder. However it does not pass this information to the hooks at all. So the hooks end up doing this:

```sh
  dir="$(git rev-parse --show-toplevel)"
  osArch=$(echo "$(uname)" | tr '[:upper:]' '[:lower:]')
  cpuArch=$(echo "$(uname -m)" | sed 's/aarch64/arm64/')

  if lefthook -h >/dev/null 2>&1
  then
    eval lefthook $@
  elif test -f "$dir/node_modules/lefthook/bin/index.js"
  then
    eval "\"$dir/node_modules/lefthook/bin/index.js\" $@"
  elif test -f "$dir/node_modules/@evilmartians/lefthook/bin/lefthook_${osArch}_${cpuArch}/lefthook"
  then
    eval "\"$dir/node_modules/@evilmartians/lefthook/bin/lefthook_${osArch}_${cpuArch}/lefthook\" $@"
  elif test -f "$dir/node_modules/@evilmartians/lefthook-installer/bin/lefthook_${osArch}_${cpuArch}/lefthook"
  then
    eval "\"$dir/node_modules/@evilmartians/lefthook-installer/bin/lefthook_${osArch}_${cpuArch}/lefthook\" $@"
  elif bundle exec lefthook -h >/dev/null 2>&1
  then
    bundle exec lefthook $@
  elif yarn lefthook -h >/dev/null 2>&1
  then
    yarn lefthook $@
  elif pnpm lefthook -h >/dev/null 2>&1
  then
    pnpm lefthook $@
  elif npx @evilmartians/lefthook -h >/dev/null 2>&1
  then
    npx @evilmartians/lefthook $@
  else
    echo "Can't find lefthook in PATH"
  fi
```

`git rev-parse --show-toplevel` only gets the root of the whole repository. nowhere do the hooks account for lefthook being installed in a deeper nested directory. this is a problem because lefthook can be part of a monorepo which is not just a top level javascript project. for example:

```
./JS-project/
./go-project/
./ios-project/
./android-project/
```

No top level package.json is desirable here, and lefthook lives inside of `./js-project/` but tells the hooks to check `./node_modules/` rather than `/JS-project/node_modules` which is where it actually lives. as a fallback condition it tries to hack using `npx lefthook` which is even worse because it retrieves some random cached version somewhere globally on the machine, which is even worse and causes version mismatches. 

### Lefthook version

1.3.3

### Possible Solution

1.  The hooks generated need to know where to look for lefthook in a monorepo. Add a directive to point to where lefthook is installed in the monorepo and propagate it to the hooks. For example `lefthook_location: "my-js-dir/`. Then pass that to the generated hooks as a param.

2. For the npx fallback, this is an unrelated bug but it also needs to be fixed. Change:
`npx @evilmartians/lefthook $@` 
to:
`npx @evilmartians/lefthook@latest $@`

The former is incorrect and will always get a globally installed random version of lefthook that was cached on the user's machine at some random point in time. The latter will always get the latest released version with bugfixes.

**Comment by mrexox** [comment]
> The hooks generated need to know where to look for lefthook in a monorepo. Add a directive to point to where lefthook is installed in the monorepo and propagate it to the hooks. For example lefthook_location: "my-js-dir/. Then pass that to the generated hooks as a param.

Hey! I guess the problem is - it doesn't look obvious that `root` parameter only changes the cwd for the commands and scripts. If I properly understand - the problem is the following:

- Your lefthook.yml lives in `./js-project/lefthook.yml`
- Your git root is in `./`
- You don't need to run hooks for other subprojects but you want a project root for lefthook to be in `js-project`

lefthook.yml is supposed to locate in the git root directory. Your case looks like a feature request.

Also things will become a bit more complicated if you have `js1-project` and `js2-project` with both having different versions of lefthook. How should the global git hooks work then? What if hooks interfere, how to resolve them? If we could find a solution I will consider adding this feature.

So, the main problem here is - for a monorepo with multiple projects we still have one set of git hooks, so we anyway need one reference for a lefthook executable.

BTW, do you have git modules? In this case things might be different but unfortunately I haven't tested this setup for a long time.

**Comment by shamilovtim** [comment]
Thanks for responding! Yeah, your description is correct. 

>  if you have js1-project and js2-project with both having different versions of lefthook. How should the global git hooks work then? What if hooks interfere, how to resolve them? If we could find a solution I will consider adding this feature.

I assume the easiest and most performant solution is to support a custom location directive in `lefthook.yml` for example `lefthook_location:`. This would avoid conflicts, since only one version of lefthook could be resolved through this directive. 

Also aside from this stuff note #2 above as well as if that worked correctly I probably never would've even noticed this. But it happened to install a bugged version of lefthook that got patched a while ago which was sitting in my `npx` cache.

**Comment by mrexox** [comment]
Hey! Now when you have `lefthook` NPM package installed in a subdirectory, and you have `root` option configured for that directory in your `lefthook.yml` (which must be in git repo root directory) – lefthook must be able to get the right executable from the `<subdirectory>/node_modules`. This feature added in this PR: https://github.com/evilmartians/lefthook/pull/616. Please, report back if everything is working or not.

**Comment by fmal** [comment]
@mrexox that is also my usecase (lefthook package installed in a subdirectory and `root` option configured for that directory), but your solution only works if you're using NPM as package manager i believe. If you are using `yarn` or `pnpm` that store downloaded packages in different location (such as `./.yarn/cache`) this solution breaks. I think that solution proposed by @shamilovtim would be more robust, where you simply provide (in `.lefthook.yml`) location/direction where the binary lives (similar to https://evilmartians.github.io/lefthook/usage/env.html#lefthook_bin but it's much easier to specify it in the config file)

**Comment by mrexox** [comment]
I see, but I guess that you won't be able to specify a path to lefthook when using with new `yarn` versions.

Probably more flexible way would be specifying a special execution command for lefthook, like `cd js-project2; yarn run lefthook`. Plain path must also work here (e.g. `./js-project2/some-unusual/location/lefthook`). What do you think about this?

I only have a concern about security. This can be potentially harmful when you allow to execute anything as a lefthook executable. But right now I don't see any other way to use lefthook installed by yarn/pnpm with PnP loader in a subdirectory.

**Comment by fmal** [comment]
yeah good point, i think that special execution command would be preferred.

Actually i think it might already be somewhat possible with use of https://evilmartians.github.io/lefthook/configuration/rc.html file, where you do `cd js-project2; export LEFTHOOK_BIN="${yarn bin lefthook}"`, but the limitation there is that it wouldn't work cross-platform i guess?

**Comment by mrexox** [comment]
Yes, the workaround with `LEFTHOOK_BIN` is quite complicated. I'll add a special `lefthook` option which accepts the command/path to lefthook executable.

**Comment by mrexox** [comment]
The feature was released with 1.10.5. Please share your feedback :)

**Comment by fmal** [comment]
@mrexox wow this is great, thanks for the quick release ❤️ I've tested it and it works the way i expect except one minor detail: `lefthook` seems to always generate `prepare-commit-msg` hook even if you don't have any commands specified in your `.lefthook.yml` config and that hook seems to be missing the `LefthookExe` block in the hook template. The result is that on each commit my `pre-commit` commands execute followed by "Can't find lefthook in PATH" (because the generated  `prepare-commit-msg` does not include the `LefthookExe` block)

**Comment by mrexox** [comment]
Yes, `prepare-commit-msg` is a legacy thing that was added to sync hooks as frequently as possible in case you change the hooks, so you don't have to run `lefthook install`. I'll reconsider this approach. 

But I hope that on a complete setup (with hooks) you don't have this issue.

**Comment by fmal** [comment]
if i explicitly define `prepare-commit-msg` in `lefthook.yml` and run `yarn lefthook install`, the generated git hook still doesn't include the `LefthookExe` block and the message is logged :(

**Comment by mrexox** [comment]
This is without the `lefthook` option configured, right?

**Comment by fmal** [comment]
With the `lefthook` option. All other hooks such as `pre-push` and `pre-commit` contain the exe block. It only seems to be the issue with `prepare-commit-msg`

**Comment by mrexox** [comment]
Oh, my bad, going to fix this in 1.10.6

**Comment by fmal** [comment]
@mrexox thank you! ❤️ now it works perfectly 💯
